src/Microsoft.ML.OnnxTransformer/OnnxTypes.cs

```diff
@@ -0,0 +1,548 @@
+
+                // Set the .NET type which corresponds to the first input argument, typeProto.
+                resultedType = typeof(IEnumerable<>).MakeGenericType(elementType);
+
+                // Create the element's caster to map IEnumerable<NamedOnnxValue> produced by ONNXRuntime to
+                // IEnumerable<elementType>.
+                var methodInfo = typeof(CastHelper).GetMethod(nameof(CastHelper.CastOnnxSequenceToIEnumerable));
+                var methodSpecialized = methodInfo.MakeGenericMethod(typeof(NamedOnnxValue), elementType);
+
+                // Use element-level caster to create sequence caster.
+                Func<NamedOnnxValue, object> caster = (NamedOnnxValue value) =>
+                {
+                    var enumerable = value.AsEnumerable<NamedOnnxValue>();
+                    return methodSpecialized.Invoke(null, new object[] { enumerable, elementCaster });
+                };
+
+                return caster;
+            }
+            else if (typeProto.ValueCase == OnnxCSharpToProtoWrapper.TypeProto.ValueOneofCase.MapType)
+            {
+                // Entering this scope means a ONNX Map (equivalent to IDictionary<>) will be produced.
+
+                var keyType = GetNativeScalarType(typeProto.MapType.KeyType);
+                var valueType = GetNativeType(typeProto.MapType.ValueType);
+
+                // The resulted type of the object returned by the caster below.
+                resultedType = typeof(IDictionary<,>).MakeGenericType(keyType, valueType);
+
+                // Create a method to convert NamedOnnxValue to IDictionary<keyValue, valueType>.
+                var asDictionaryMethodInfo = typeof(NamedOnnxValue).GetMethod(nameof(NamedOnnxValue.AsDictionary));
+                var asDictionaryMethod = asDictionaryMethodInfo.MakeGenericMethod(keyType, valueType);
+
+                // Create a caster to convert NamedOnnxValue to IDictionary<keyValue, valueType>.
+                Func<NamedOnnxValue, object> caster = (NamedOnnxValue value) =>
+                {
+                    return asDictionaryMethod.Invoke(value, new object[] { });
+                };
+
+                return caster;
+            }
+            else
+                throw Contracts.ExceptParamValue(typeProto, nameof(typeProto), $"Unsupported ONNX variable type {typeProto}");
+        }
+    }
+
+    /// <summary>
+    /// The corresponding <see cref="DataViewSchema.Column.Type"/> of ONNX's sequence type in <see cref="IDataView"/>'s type system.
+    /// In other words, if an ONNX model produces a sequence, a column in <see cref="IDataView"/> may be typed to <see cref="OnnxSequenceType"/>.
+    /// Its underlying type is <see cref="IEnumerable{T}"/>, where the generic type "T" is the input argument of
+    /// <see cref="OnnxSequenceType.OnnxSequenceType(Type)"/>.
+    /// </summary>
+    public sealed class OnnxSequenceType : StructuredDataViewType
```

Comment:
Can you put top-level classes in their own file? One class per file is usually a good recommendation.

Here, the first class in the file is "internal", which makes it look like there are no public APIs being added.  But the public APIs are later in the file.

Reply:
No problem. Btw, I still put "...DataViewType" and its "...DataViewTypeAttribute" in the same file for emphasizing that they are dual of each other.

---
In reply to: [298374249](https://github.com/dotnet/machinelearning/pull/3881#discussion_r298374249) [](ancestors = 298374249)